### PR TITLE
Optimize performance of softmax_fwd when axis!=-1

### PR DESCRIPTION
**Pull request by ZzSean**
### PR types

Performance optimization
### PR changes

OPs
### Describe

Optimize performance of softmax_fwd when axis!=-1
| case                         | pytorch | paddle 优化前 | diff          | paddle 优化后 | diff          | 加速比 |
| ---------------------------- | ------- | ------------- | ------------- | ------------- | ------------- | ------ |
| fp32,[128,128,16,16]  axis:0 | 0.23059 | 0.18179       | 优于 (26.84%) | 0.07534       | 优于 (2.06x)  | 2.41   |
| fp16,[128,128,16,16] axis:0  | 0.20328 | 0.14136       | 优于 (43.80%) | 0.06305       | 优于 (2.22x)  | 2.24   |
| fp32,[512,896,4,12] axis:1   | 0.76943 | 1.22761       | 差于 (59.55%) | 0.56893       | 优于 (35.24%) | 2.16   |
| fp16,[512,896,4,12] axis:1   | 0.62722 | 1.02623       | 差于 (63.62%) | 0.41894       | 优于 (49.72%) | 2.45   |

### CODE REVIEW

File: paddle/fluid/operators/softmax_cudnn_op.cu.h [hunk 1]
```diff
@@ -444,6 +497,109 @@ void SwitchWarpSoftmaxBackward(const int blocks, const dim3 threads,
...
+
+static void GetBlockDim(int mid_dim, int low_dim, dim3* block) {
+  constexpr int max_num_threads = 1024;
+  int block_x = 1 << log2_ceil(low_dim);
+  int block_y = 1 << log2_ceil(mid_dim);
+  block->x = std::min(block_x, 32);
+  block->y = std::min(block_y, static_cast<int>(max_num_threads / block->x));
+  block->x = std::min(block_x, static_cast<int>(max_num_threads / block->y));
+}
+
```
Comment (Liu-xiandong) [inline paddle/fluid/operators/softmax_cudnn_op.cu.h:534]: 想了解一下，这个优化思路是什么呢？针对不同的tensor shape来设置不同的block和thread吗？

Comment (ZzSean) [reply paddle/fluid/operators/softmax_cudnn_op.cu.h:534]: 优化思路跟reduce类似，主要是两个：
1）当规约的维度不是最低维时，让每个warp还可以保证合并访存；
2）使用二维的 block 和 grid，尽可能地保证 thread 的并行度和资源利用最大化

File: paddle/fluid/operators/softmax_cudnn_op.cu.h [hunk 2]
```diff
@@ -444,6 +497,109 @@ void SwitchWarpSoftmaxBackward(const int blocks, const dim3 threads,
...
+  int grid_x = (low_dim + block.x - 1) / block.x;
+  grid_x = std::min(grid_x, max_num_blocks);
+  int grid_y = (max_num_blocks + grid_x - 1) / grid_x;
+  grid_y = std::min(grid_y, high_dim);
+  grid->x = grid_x;
+  grid->y = grid_y;
+}
+
+static void GetBlockDim(int mid_dim, int low_dim, dim3* block) {
+  constexpr int max_num_threads = 1024;
```
Comment (AnnaTrainingG) [inline paddle/fluid/operators/softmax_cudnn_op.cu.h:526]: 当前kernel是否需要支持HIP呢？ HIP的最大线程数量应该 <= 256

File: paddle/fluid/operators/softmax_cudnn_op.cu.h [hunk 3]
```diff
@@ -444,6 +497,109 @@ void SwitchWarpSoftmaxBackward(const int blocks, const dim3 threads,
...
+         low_id += blockDim.x * gridDim.x) {
+      const int input_offset = high_id * high_stride + low_id;
+
+      // 1. reduce max
+      AccT max_value = -std::numeric_limits<AccT>::infinity();
+      AccT value = -std::numeric_limits<AccT>::infinity();
+      for (int mid_id = threadIdx.y; mid_id < mid_dim; mid_id += blockDim.y) {
+        value = static_cast<AccT>(input[input_offset + mid_id * mid_stride]);
+        max_value = kps::MaxFunctor<AccT>()(max_value, value);
+      }
```
Comment (niuliling123) [inline paddle/fluid/operators/softmax_cudnn_op.cu.h:556]: 为什么使用的是（xi-max)的形式，这样会不会多求做了一次max
https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/Softmax_cn.html#softmax
这两种在性能上应该会有差别吧

File: paddle/fluid/operators/softmax_cudnn_op.cu.h (hunk 2 above)
Comment (ZzSean) [reply paddle/fluid/operators/softmax_cudnn_op.cu.h:526]: 是需要的，这里加个宏

**Review by Liu-xiandong — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/softmax_cudnn_op.cu.h [hunk 4]
```diff
@@ -444,6 +497,113 @@ void SwitchWarpSoftmaxBackward(const int blocks, const dim3 threads,
...
+      if (blockDim.y > 1) {
+        kps::Reduce<AccT, 1, 1, 1, kps::AddFunctor<AccT>, kMode::kGlobalMode>(
+            &sum, &sum, kps::AddFunctor<AccT>(), false);
+      }
+
+      // 3. (log)softmax
+      Functor<AccT, T> functor(max_value, sum);
+      for (int mid_id = threadIdx.y; mid_id < mid_dim; mid_id += blockDim.y) {
+        int data_offset = input_offset + mid_id * mid_stride;
+        output[data_offset] = functor(static_cast<AccT>(input[data_offset]));
```
Comment (AshburnLee) [inline paddle/fluid/operators/softmax_cudnn_op.cu.h:581]: 三步骤中都有对`input`的读，且`data_offset`一样，是否可优化只读一次?

Comment (AshburnLee) [inline paddle/fluid/operators/softmax_cudnn_op.cu.h:581]: 是否能用`kp`中的`ReadDataBc`或`ReadData`实现读？

File: paddle/fluid/operators/softmax_cudnn_op.cu.h [hunk 5]
```diff
@@ -444,6 +497,113 @@ void SwitchWarpSoftmaxBackward(const int blocks, const dim3 threads,
...
+      // 1. reduce max
+      AccT max_value = -std::numeric_limits<AccT>::infinity();
+      AccT value = -std::numeric_limits<AccT>::infinity();
+      for (int mid_id = threadIdx.y; mid_id < mid_dim; mid_id += blockDim.y) {
+        value = static_cast<AccT>(input[input_offset + mid_id * mid_stride]);
+        max_value = kps::MaxFunctor<AccT>()(max_value, value);
+      }
+
+      if (blockDim.y > 1) {
+        kps::Reduce<AccT, 1, 1, 1, kps::MaxFunctor<AccT>, kMode::kGlobalMode>(
```
Comment (AshburnLee) [inline paddle/fluid/operators/softmax_cudnn_op.cu.h:562]: 这种情况下，也能用`BlockYReduce`吧，`kps::Reduce`为什么会比传统的block reduce快呢？

File: paddle/fluid/operators/softmax_cudnn_op.cu.h (hunk 4 above)
Comment (ZzSean) [reply paddle/fluid/operators/softmax_cudnn_op.cu.h:581]: 第一次是求max，第二次是求sum，循环次数较多，没办法用那么多的寄存器把每次结果都保存

Comment (ZzSean) [reply paddle/fluid/operators/softmax_cudnn_op.cu.h:581]: 这里没有向量化，用kp实现可以，但是代码会变得复杂

File: paddle/fluid/operators/softmax_cudnn_op.cu.h (hunk 5 above)
Comment (ZzSean) [reply paddle/fluid/operators/softmax_cudnn_op.cu.h:562]: 这里用的就是BlockYReduce啊

**Review by AnnaTrainingG — APPROVED** [review]
LGTM
